src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs

```diff
@@ -344,141 +354,154 @@ private static unsafe bool AllowHyphenDuringParsing(NumberFormatInfo info)
-                    // Eat whitespace unless we've found a sign which isn't followed by a currency symbol.
-                    // "-Kr 1231.47" is legal but "- 1231.47" is not.
-                    if (!IsWhite(ch) || (options & NumberStyles.AllowLeadingWhite) == 0 || ((state & StateSign) != 0 && ((state & StateCurrency) == 0 && numfmt.NumberNegativePattern != 2)))
-                    {
-                        if ((((options & NumberStyles.AllowLeadingSign) != 0) && (state & StateSign) == 0) && ((next = MatchChars(p, strEnd, numfmt.PositiveSign)) != null || ((next = MatchNegativeSignChars(p, strEnd, numfmt.NegativeSign, allowHyphenDuringParsing)) != null && (number.sign = true))))
-                        {
-                            state |= StateSign;
-                            p = next - 1;
-                        }
-                        else if (ch == '(' && ((options & NumberStyles.AllowParentheses) != 0) && ((state & StateSign) == 0))
-                        {
-                            state |= StateSign | StateParens;
-                            number.sign = true;
-                        }
-                        else if (currSymbol != null && (next = MatchChars(p, strEnd, currSymbol)) != null)
-                        {
-                            state |= StateCurrency;
-                            currSymbol = null;
+                    returnValue = new FormattingSpecification(numfmt.CurrencySymbol,
+                        numfmt.CurrencyDecimalSeparator,
+                        numfmt.CurrencyGroupSeparator);
+                }
+                else
+                {
+                    returnValue = new FormattingSpecification(null,
+                        numfmt.NumberDecimalSeparator,
+                        numfmt.NumberGroupSeparator);
+                }
 
-                            // We already found the currency symbol. There should not be more currency symbols. Set
-                            // currSymbol to NULL so that we won't search it again in the later code path.
-                            p = next - 1;
-                        }
-                        else
-                        {
-                            break;
-                        }
-                    }
-                    ch = ++p < strEnd ? *p : '\0';
+                return returnValue;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static unsafe char* MatchNegativeSignChars(char* p, char* pEnd, string negativeSign, bool allowHyphenDuringParsing)
+            {
+                char *ret = MatchChars(p, pEnd, negativeSign);
+                if (ret == null && allowHyphenDuringParsing && p < pEnd && *p == '-')
+                {
+                    ret = p + 1;
                 }
 
-                int digCount = 0;
-                int digEnd = 0;
+                return ret;
+            }
+
+            private static unsafe StringBuilder ParseCommon(
```

Comment:
This needs testing/profiling to show that the additional helper functions are getting inlined and not impacting the hot path.

Likewise, this is going to end up checking `IsDecimalChar(ch)` twice now, since `IsHexadecimalChar` tests it itself.